src/cls/queue/cls_queue.cc

```diff
@@ -0,0 +1,1257 @@
+{
+  // read the head
+  bufferlist bl_head;
+  int ret = cls_cxx_read2(hctx, 0, QUEUE_HEAD_SIZE, &bl_head, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    return ret;
+  }
+  cls_queue_head head;
+  auto iter = bl_head.cbegin();
+  try {
+    decode(head, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_queue_write_urgent_data: failed to decode entry %s\n", bl_head.c_str());
+    return -EINVAL;
+  }
+
+  CLS_LOG(1, "INFO: cls_queue_write_urgent_data: tail offset %lu\n", head.tail);
+
+  auto in_iter = in->cbegin();
+
+  cls_queue_write_urgent_data_op op;
+  try {
+    decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_queue_write_urgent_data: failed to decode input data\n");
+    return -EINVAL;
+  }
+  //Write urgent data
+  head.has_urgent_data = op.has_urgent_data;
+  head.bl_urgent_data = op.bl_urgent_data;
+
+  //Write head back
+  bl_head.clear();
+  encode(head, bl_head);
+  CLS_LOG(1, "INFO: cls_queue_write_urgent_data: Writing head of size: %u\n", bl_head.length());
+  ret = cls_cxx_write2(hctx, 0, bl_head.length(), &bl_head, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    CLS_LOG(1, "INFO: cls_queue_write_urgent_data: Writing head returned error: %d \n", ret);
+    return ret;
+  }
+
+  return 0;
+}
+
+static int cls_queue_can_urgent_data_fit(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  bool can_fit = true;
+
+  // read the head
+  bufferlist bl_head;
+  int ret = cls_cxx_read2(hctx, 0, QUEUE_HEAD_SIZE, &bl_head, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    return ret;
+  }
+  cls_queue_head head;
+  auto iter = bl_head.cbegin();
+  try {
+    decode(head, iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_queue_write_urgent_data: failed to decode entry %s\n", bl_head.c_str());
+    return -EINVAL;
+  }
+
+  head.has_urgent_data = true;
+  head.bl_urgent_data = *in;
+
+  bl_head.clear();
+  encode(head, bl_head);
+
+  if(bl_head.length() > QUEUE_HEAD_SIZE) {
+    can_fit = true;
+  }
+
+  encode(can_fit, *out);
+
+  return 0;
+}
+
+static int cls_gc_create_queue(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  auto in_iter = in->cbegin();
+
+  cls_gc_create_queue_op op;
+  try {
+    decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_gc_create_queue: failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  cls_create_queue_op create_op;
+
+  if (op.num_urgent_data_entries > 0) {
+    std::unordered_map<string,ceph::real_time> urgent_data_map;
+    urgent_data_map.reserve(op.num_urgent_data_entries);
+    encode(urgent_data_map, create_op.head.bl_urgent_data);
+  }
+
+  create_op.head.size = op.size;
+
+  in->clear();
+  encode(create_op, *in);
+
+  return cls_create_queue(hctx, in, out);
+}
+
+static int cls_gc_enqueue(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  auto in_iter = in->cbegin();
+
+  cls_rgw_gc_set_entry_op op;
```

Comment:
I thought that processing entries returned by queue list for GC list and remove will be faster on the OSD side instead of sending them back to the client (which I think is part of RGW). I am not sure exactly where to put these GC interfaces so I made them part of the queue interface.

Reply:
@cbodley : One more reason why I made gc_list and gc_remove a part of cls_queue (instead of having them in the cls_queue_client) is because all steps should complete as part of a single transaction (there are many sub-steps related to urgent data, and then checking for deferred entries etc). These can be moved out of cls_queue and placed in a separate cls module as you were suggesting yesterday. What do you think?

Reply:
Can I just move out the cls gc queue interfaces in another file say cls_gc_queue.cc and then make the methods in cls_queue.cc non-static and put them in a header file and include that in cls_gc_queue.cc?